wsmaster/che-core-api-auth-github/src/main/java/org/eclipse/che/security/oauth/GitHubOAuthAuthenticatorProvider.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2012-2021 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.security.oauth;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Objects;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+import org.eclipse.che.api.auth.shared.dto.OAuthToken;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.security.oauth.shared.User;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+public class GitHubOAuthAuthenticatorProvider implements Provider<OAuthAuthenticator> {
+  private static final Logger LOG = LoggerFactory.getLogger(GitHubOAuthAuthenticatorProvider.class);
+  public static final String GITHUB_CLIENT_ID_PATH = "/che-conf/oauth/github/id";
+  public static final String GITHUB_CLIENT_SECRET_PATH = "/che-conf/oauth/github/secret";
+
+  private final OAuthAuthenticator authenticator;
+
+  @Inject
+  public GitHubOAuthAuthenticatorProvider(
+      @Nullable @Named("che.oauth.github.redirecturis") String[] redirectUris,
+      @Nullable @Named("che.oauth.github.authuri") String authUri,
+      @Nullable @Named("che.oauth.github.tokenuri") String tokenUri)
+      throws IOException {
+    authenticator =
+        getOAuthAuthenticator(
+            redirectUris, authUri, tokenUri, GITHUB_CLIENT_ID_PATH, GITHUB_CLIENT_SECRET_PATH);
+  }
+
+  @Override
+  public OAuthAuthenticator get() {
+    return authenticator;
+  }
+
+  @VisibleForTesting
+  OAuthAuthenticator getOAuthAuthenticator(
+      String[] redirectUris,
+      String authUri,
+      String tokenUri,
+      String clientIdPath,
+      String clientSecretPath)
+      throws IOException {
+
+    if (isNullOrEmpty(authUri)
+        || isNullOrEmpty(tokenUri)
+        || Objects.isNull(redirectUris)
+        || redirectUris.length == 0) {
+      LOG.debug(
+          "URIs for GitHub OAuth authentication are missing or empty. Make sure your configuration is correct.");
+      return new NoopOAuthAuthenticator();
+    }
+
+    String clientId, clientSecret;
+    try {
+      clientId = Files.readString(Path.of(clientIdPath));
+      clientSecret = Files.readString(Path.of(clientSecretPath));
+    } catch (IOException e) {
+      LOG.debug(
```

Comment:
I don't think that is a good idea to catch IOException here. AFAIK BitbucketServerOAuthAuthenticatorProvider just rethrows it for the caller. I think it's ok to fail the container here. `+` it will make the code a bit cleaner.

Reply:
But this will happen everytime the github-oauth-config secret is not present (or the credentials are not mounted). So for users that didnt setup github oauth, when che is deployed, there will be an error output in the logs from the guice container about files not exising. And that I think seems confusing.

Reply:
Do you mean IOException ?

Reply:
my point is if clientIdPath, clientSecretPath is set and not null and if during reading we get IOException  - then we should not silently ignore that by setting up Noop implementation.

Reply:
Ah yes good point. But still most of the time the error would come from just not configuring the github-oauth. So would it make sense to catch FileNotFoundException but propagate other IOExceptions, in your opinion?

Reply:
What about make it similar to https://github.com/eclipse-che/che-server/blob/388a5183be0366a5d48c3845c35c618b5dbaeac9/wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth1/BitbucketServerOAuthAuthenticatorProvider.java#L54-L65 ?